dlt/extract/pipe.py

```diff
@@ -718,21 +720,22 @@ def _get_source_item_current(self) -> ResolvablePipeItem:
             raise ResourceExtractionError(pipe.name, gen, str(ex), "generator") from ex
 
     def _get_source_item_round_robin(self) -> ResolvablePipeItem:
-        # no more sources to iterate
         sources_count = len(self._sources)
+        # no more sources to iterate
         if sources_count == 0:
             return None
         # if there are currently more sources than added initially, we need to process the new ones first
         if sources_count > self._initial_sources_count:
             return self._get_source_item_current()
         try:
-            # get items from last added iterator, this makes the overall Pipe as close to FIFO as possible
-            self._round_robin_index = (self._round_robin_index + 1) % sources_count
-            gen, step, pipe, meta = self._sources[self._round_robin_index]
             # print(f"got {pipe.name} {pipe._pipe_id}")
             # register current pipe name during the execution of gen
-            set_current_pipe_name(pipe.name)
-            item = next(gen)
+            item = None
+            while item is None:
+                self._round_robin_index = (self._round_robin_index + 1) % sources_count
+                gen, step, pipe, meta = self._sources[self._round_robin_index]
+                set_current_pipe_name(pipe.name)
```

Comment:
can this go below the loop, or is this somehow used for exceptions that occur during the next call that follows?

Reply:
it should be set immediately when a new potential item is taken from _sources. so it looks good